File: ts-packages/src/lib/methods.ts
import { db } from "@/lib/db";
import { and, desc, eq } from "drizzle-orm";
import {
  conversationsTable,
  ConversationStatus,
  messagesTable,
  pdfDocumentsTable,
  productReviewsTable,
  productsTable,
  usersTable,
} from "./schema";
import type {
  CreateConversationResult,
  CreateMessageResult,
  CreatePdfDocumentResult,
  CreateProductResult,
  CreateProductReviewResult,
  CreateUserResult,
  GetConversationByIdResult,
  GetConversationsResult,
  GetMessagesResult,
  GetPdfDocumentsResult,
  GetProductByIdResult,
  GetProductReviewsResult,
  GetProductsResult,
  GetUserByIdResult,
  UpdateConversationResult,
} from "./types";

type ConversationStatusType =
  (typeof ConversationStatus)[keyof typeof ConversationStatus];

// ---------------------------------------------------------------------------
// Products
// ---------------------------------------------------------------------------

export async function getProducts(filters?: {
  query?: string;
  company?: string;
  category?: string;
  sortBy?: "createdAt" | "price" | "name";
  take?: number;
  ratings?: number;
  noOfRatings?: number;
}): GetProductsResult {
  if (!filters) return await db.select().from(productsTable).limit(10);

  const sortByMap = {
    createdAt: desc(productsTable.createdAt),
    price: desc(productsTable.price),
    name: desc(productsTable.name),
  };
  const conditions = [];
  let sortBy = sortByMap[filters.sortBy ?? "createdAt"];
  let where;
  if (filters?.query) {
    conditions.push(eq(productsTable.query, filters.query));
  }

  if (filters?.company) {
    conditions.push(eq(productsTable.company, filters.company));
  }

  if (filters?.category) {
    conditions.push(eq(productsTable.category, filters.category));
  }

  if (filters?.ratings) {
    conditions.push(eq(productsTable.ratings, filters.ratings));
  }

  if (filters?.noOfRatings) {
    conditions.push(eq(productsTable.noOfRatings, filters.noOfRatings));
  }

  let take = filters.take ?? 10;

  const query = db
    .select()
    .from(productsTable)
    .where(conditions.length ? and(...conditions) : undefined)
    .orderBy(sortByMap[filters?.sortBy ?? "createdAt"])
    .limit(take);
  return await query;
}

export async function getProductById(id: string): GetProductByIdResult {
  const rows = await db
    .select()
    .from(productsTable)
    .where(eq(productsTable.id, id));
  return rows[0];
}

export async function getProductReviews(
  productId: string,
): GetProductReviewsResult {
  return db
    .select()
    .from(productReviewsTable)
    .where(eq(productReviewsTable.productId, productId));
}

// ---------------------------------------------------------------------------
// Users
// ---------------------------------------------------------------------------

export async function getUserById(id: string): GetUserByIdResult {
  const rows = await db.select().from(usersTable).where(eq(usersTable.id, id));
  return rows[0];
}

// ---------------------------------------------------------------------------
// Conversations
// ---------------------------------------------------------------------------

export async function getConversations(
  userId?: string,
): GetConversationsResult {
  if (userId) {
    return db
      .select()
      .from(conversationsTable)
      .where(eq(conversationsTable.userId, userId));
  }
  return db.select().from(conversationsTable);
}

export async function getConversationById(
  id: string,
): GetConversationByIdResult {
  const rows = await db
    .select()
    .from(conversationsTable)
    .where(eq(conversationsTable.id, id));
  return rows[0];
}

export async function deleteConversation(id: string) {
  const rows = await db
    .delete(conversationsTable)
    .where(eq(conversationsTable.id, id));
  return rows;
}
// ---------------------------------------------------------------------------
// Messages
// ---------------------------------------------------------------------------

export async function getMessages(conversationId: string): GetMessagesResult {
  return db
    .select()
    .from(messagesTable)
    .where(eq(messagesTable.conversationId, conversationId));
}

// ---------------------------------------------------------------------------
// PDF Documents
// ---------------------------------------------------------------------------

export async function getPdfDocuments(
  conversationId: string,
): GetPdfDocumentsResult {
  return db
    .select()
    .from(pdfDocumentsTable)
    .where(eq(pdfDocumentsTable.conversationId, conversationId))
    .orderBy(desc(pdfDocumentsTable.updatedAt));
}

// ---------------------------------------------------------------------------
// Create / Update — Products
// ---------------------------------------------------------------------------

export async function createProduct(data: {
  name?: string;
  url?: string;
  price?: number;
  query?: string;
  company?: string;
  category?: string;
  ratings?: number;
  noOfRatings?: number;
  reviews?: string[];
}): CreateProductResult {
  const [row] = await db.insert(productsTable).values(data).returning();
  if (row) {
    for (const review of data.reviews || []) {
      await createProductReview({ productId: row.id, reviewText: review });
    }
  }
  return row;
}

export async function createProductReview(data: {
  productId: string;
  reviewText?: string;
}): CreateProductReviewResult {
  const [row] = await db.insert(productReviewsTable).values(data).returning();
  return row;
}

// ---------------------------------------------------------------------------
// Create — Users
// ---------------------------------------------------------------------------

export async function createUser(data: {
  id: string;
  email: string;
}): CreateUserResult {
  const [row] = await db.insert(usersTable).values(data).returning();
  return row;
}

// ---------------------------------------------------------------------------
// Create / Update — Conversations
// ---------------------------------------------------------------------------

export async function createConversation(data: {
  query: string;
  userId: string;
  status?: ConversationStatusType;
  openai_convId?: string;
}): CreateConversationResult {
  const [row] = await db
    .insert(conversationsTable)
    .values({ ...data, status: data.status ?? ConversationStatus.PENDING })
    .returning();
  return row;
}

export async function updateConversation(data: {
  id: string;
  status: ConversationStatusType;
  userId?: string;
}): UpdateConversationResult {
  const conditions = [eq(conversationsTable.id, data.id)];
  const [row] = await db
    .update(conversationsTable)
    .set({ status: data.status })
    .where(conditions.length === 1 ? conditions[0] : conditions[0])
    .returning();
  return row;
}

// ---------------------------------------------------------------------------
// Create — Messages
// ---------------------------------------------------------------------------

export async function createMessage(data: {
  conversationId: string;
  userId?: string;
  content: string;
  role?: "user" | "assistant";
}): CreateMessageResult {
  const [row] = await db
    .insert(messagesTable)
    .values({ ...data, role: data.role ?? "user" })
    .returning();
  return row;
}

// ---------------------------------------------------------------------------
// Create — PDF Documents
// ---------------------------------------------------------------------------

export async function createPdfDocument(data: {
  conversationId: string;
  fileName?: string;
  filePath?: string;
}): CreatePdfDocumentResult {
  const [row] = await db.insert(pdfDocumentsTable).values(data).returning();
  return row;
}

File: py_packages/py_packages/lib/methods.py
import datetime
from typing import Literal, Optional
import uuid
from sqlalchemy.orm import selectinload
from sqlmodel import Session, select

ConversationStatus = Literal[
    "pending", "in_progress", "completed", "scraping", "generation", "embedding"
]

from .models import (
    Conversations,
    Messages,
    PdfDocuments,
    ProductReviews,
    Products,
    Users,
)


# ---------------------------------------------------------------------------
# Products
# ---------------------------------------------------------------------------


def get_products(
    session: Session,
    query: Optional[str] = None,
    company: Optional[str] = None,
    category: Optional[str] = None,
) -> list[Products]:
    statement = select(Products).options(selectinload(Products.product_reviews))
    if query is not None:
        statement = statement.where(Products.query == query)
    if company is not None:
        statement = statement.where(Products.company == company)
    if category is not None:
        statement = statement.where(Products.category == category)
    return session.exec(statement).all()


def get_product_by_id(session: Session, product_id: uuid.UUID) -> Optional[Products]:
    return session.get(Products, product_id)


def get_product_reviews(session: Session, product_id: uuid.UUID) -> list[ProductReviews]:
    statement = select(ProductReviews).where(ProductReviews.product_id == product_id)
    return session.exec(statement).all()


# ---------------------------------------------------------------------------
# Users
# ---------------------------------------------------------------------------


def get_user_by_id(session: Session, user_id: uuid.UUID) -> Optional[Users]:
    return session.get(Users, user_id)


def get_user_by_email(session: Session, email: str) -> Optional[Users]:
    statement = select(Users).where(Users.email == email)
    return session.exec(statement).first()


# ---------------------------------------------------------------------------
# Conversations
# ---------------------------------------------------------------------------


def get_conversations(
    session: Session,
    user_id: Optional[uuid.UUID] = None,
) -> list[Conversations]:
    statement = select(Conversations)
    if user_id is not None:
        statement = statement.where(Conversations.user_id == user_id)
    return session.exec(statement).all()


def get_conversation_by_id(
    session: Session, conversation_id: uuid.UUID
) -> Optional[Conversations]:
    return session.get(Conversations, conversation_id)


# ---------------------------------------------------------------------------
# Messages
# ---------------------------------------------------------------------------


def get_messages(session: Session, conversation_id: uuid.UUID) -> list[Messages]:
    statement = select(Messages).where(Messages.conversation_id == conversation_id)
    return session.exec(statement).all()


# ---------------------------------------------------------------------------
# PDF Documents
# ---------------------------------------------------------------------------


def get_pdf_documents(
    session: Session, conversation_id: uuid.UUID
) -> list[PdfDocuments]:
    statement = select(PdfDocuments).where(
        PdfDocuments.conversation_id == conversation_id
    )
    return session.exec(statement).all()


# ---------------------------------------------------------------------------
# Create / Update — Products
# ---------------------------------------------------------------------------


def create_product(
    session: Session,
    name: Optional[str] = None,
    url: Optional[str] = None,
    price: Optional[float] = None,
    query: Optional[str] = None,
    company: Optional[str] = None,
    category: Optional[str] = None,
) -> Products:
    product = Products(name=name, url=url, price=price, query=query, company=company, category=category)
    session.add(product)
    session.commit()
    session.refresh(product)
    return product


def create_product_review(
    session: Session,
    product_id: uuid.UUID,
    review_text: Optional[str] = None,
) -> ProductReviews:
    review = ProductReviews(product_id=product_id, review_text=review_text)
    session.add(review)
    session.commit()
    session.refresh(review)
    return review


# ---------------------------------------------------------------------------
# Create — Users
# ---------------------------------------------------------------------------


def create_user(
    session: Session,
    email: str,
    name: Optional[str] = None,
) -> Users:
    user = Users(email=email, name=name)
    session.add(user)
    session.commit()
    session.refresh(user)
    return user


# ---------------------------------------------------------------------------
# Create / Update — Conversations
# ---------------------------------------------------------------------------


def create_conversation(
    session: Session,
    query: str,
    user_id: uuid.UUID,
    status: ConversationStatus = "pending",
    openai_conv_id: Optional[str] = None,
) -> Conversations:
    conversation = Conversations(
        query=query, user_id=user_id, status=status, openai_conv_id=openai_conv_id
    )
    session.add(conversation)
    session.commit()
    session.refresh(conversation)
    return conversation


def update_conversation(
    session: Session,
    conversation_id: uuid.UUID,
    status: ConversationStatus,
) -> Optional[Conversations]:
    conversation = session.get(Conversations, conversation_id)
    if not conversation:
        return None
    conversation.status = status
    session.add(conversation)
    session.commit()
    session.refresh(conversation)
    return conversation


# ---------------------------------------------------------------------------
# Create — Messages
# ---------------------------------------------------------------------------


def create_message(
    session: Session,
    conversation_id: uuid.UUID,
    content: str,
    user_id: Optional[uuid.UUID] = None,
    role: Literal["user", "assistant"] = "user",
    updated_at: Optional[datetime.datetime] = datetime.datetime.now(),
) -> Messages:
    message = Messages(
        conversation_id=conversation_id, user_id=user_id, content=content, role=role,
        updated_at=updated_at
    )
    session.add(message)
    session.commit()
    session.refresh(message)
    return message


# ---------------------------------------------------------------------------
# Create — PDF Documents
# ---------------------------------------------------------------------------


def create_pdf_document(
    session: Session,
    conversation_id: uuid.UUID,
    file_name: Optional[str] = None,
    file_path: Optional[str] = None,
    updated_at: Optional[datetime.datetime] = datetime.datetime.now(),
) -> PdfDocuments:
    doc = PdfDocuments(
        conversation_id=conversation_id, file_name=file_name, file_path=file_path,
        updated_at=updated_at
    )
    session.add(doc)
    session.commit()
    session.refresh(doc)
    return doc


def get_pdf_documents(session, conversation_id):
    statement = select(PdfDocuments).where(
        PdfDocuments.conversation_id == conversation_id
    )   
    return session.exec(statement).first()
